### DISCUSSION TITLE
3rd Party App (TeslaMate) interferes with car systems

### DISCUSSION THREAD

**Discussion by banksy-git** [teslamate-org/teslamate#1833] [General]
I have happily used TeslaMate for a little over a year and have enjoyed the stats it provides. Having an API, albeit not officially supported, is a great part of Tesla ownership and, like many of you I'm sure, it certainly it tickles my inner geek!

Last week I put my car into the shop because a number of faults developed :

1. SOS service required message displayed.
2. Reverse camera didn't work.
3. Autopilot and cruise control didn't work
4. GPS map didn't update.
5. Visualisations were not showing.

I tried the two steering-wheel button (and brake pedal) reboot but that did not resolve the fault so I booked it into service.

The Tesla service department just got back to me. After thoroughly checking the hardware components and related wiring they said **that the fault was, ultimately, caused by a 3rd party app. (i.e. TeslaMate - it's the only one I use.)**

Around time of the fault, a component in the car (I will hopefully get details from their technical write up tomorrow), as part of its normal operation, rebooted. The Tesla tech's theory - supported by internal logs taken by Tesla - is that the reboot of this component coincided with TeslaMate querying the car. The two actions together caused a bus contention that was not resolved and that subsequently prevented all the components on that bus from communicating; hence the large number of apparent faults.

Additionally, the car then did not sleep for 4 days solid, which Tesla informs me was also due to the actions of the 3rd party app. (Oddly the TeslaMate logs I have don't show any abnormal awake periods but I have no reason to doubt the Tesla techs.)

Another clue is that, around the time of the fault, Tesla revoked the API refresh token that Teslamate was using to query the car. (I only noticed that today because Tesla told me about the 3rd party app issue so I went to hunt logs!) I'll try to line up all the timestamps later if people are interested.

From chatting with the service techs, it seems Tesla's position is that the API and Tesla software ecosystem is a coherent whole and this is a good example why 3rd party API usage is not supported! Nonetheless, as a principle of hardware/firmware/software design, of which I have _significant_ experience, I consider that behaviour a bug in the Tesla firmware. A web API should not be able to cause such faults.

So, has anyone else experienced such a fault? Or other weirdness. If you do... it might be the 3rd party app!

**Comment by ngardiner** [comment]
Tesla regularly blame / require disabling of 3rd party interfaces. Because they are not sanctioned by Tesla, they can get away with it, and it then plays into the narrative that 3rd party apps are bad.

All TeslaMate does is query the same endpoints as the Tesla app does. People have only ever learned the API endpoints that have been exposed through analysis of the app, since no Tesla documentation exists. Even if there was bus contention caused by a reboot of a component at the same time as an API query, the app would have left the vehicle in the same state, how would it have known not to query? 

Ultimately, this is a lazy excuse from Tesla that they are able to give because they don't sanction 3rd party apps. We are in the wrong for using the API to add functionality to our vehicles. If you don't want to be blamed again in the future, you can stop using 3rd party apps, it is up to you.

Next time, I recommend you disable TeslaMate before the service so Tesla don't just blame it and close off the issue.

**Reply by banksy-git** [reply]
I agree.

To be fair to the Tesla techs, I believe they did investigate thoroughly; testing the wiring harness and the components. Additionally, you'd think with an "unsanctioned 3rd party app" conclusion they'd be within their rights to charge me for their time - but they didn't - which I'm grateful for.

Ultimately, if the Tesla theory is correct, there's a bug in the car firmware(s). The firmware should be resilient to external queries that can land at any time and a fault like this should not occur. Indeed, I am skeptical that this was the cause of the faults. Time will tell!
